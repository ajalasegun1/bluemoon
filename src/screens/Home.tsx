import {FlatList, StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import React, {FC, useCallback, useContext, useState} from 'react';
import {SafeAreaView} from 'react-native-safe-area-context';
import CustomButton from '../components/CustomButton';
import handleLogout from '../helper/logout';
import {HomeScreenProps} from '../navigation/types';
import AppContext from '../context/AppContext';
import AddIcon from '../assets/svgs/add.svg';
import {Inventory} from '../context/types';
import {COLORS} from '../constants/theme';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {useFocusEffect} from '@react-navigation/native';

const Home: FC<HomeScreenProps> = ({navigation}) => {
  const {setUser, user} = useContext(AppContext);
  const [data, setData] = useState<Inventory[] | null>([]);

  useFocusEffect(
    useCallback(() => {
      fetchInventory();
    }, []),
  );

  const logout = async () => {
    try {
      const success = await handleLogout();
      if (success) {
        setUser(null);
        return true;
      }
      return false;
    } catch (error) {
      console.log({error});
    }
  };
  const goToAddScreen = () => navigation.navigate('AddInventory');

  const fetchInventory = useCallback(async () => {
    // From Async Storage
    const data = await AsyncStorage.getItem('inventory');
    const inventory = JSON.parse(data as any);
    if (inventory && inventory.length > 0 && user?.email) {
      const filter = inventory.filter(
        (item: Inventory) => item.email === user.email,
      );
      setData(filter);
    } else {
      setData(null);
    }
  }, []);

  const renderItem = ({item, index}: {item: Inventory; index: number}) => (
    <TouchableOpacity
      style={styles.tableRow}
      onPress={() => goToEdit(item.name)}
      key={index.toString()}
      testID="data-row">
      <View style={styles.cellStyle2}>
        <Text>{item.name}</Text>
      </View>
      <View style={styles.cellStyle2}>
        <Text>{item.desc}</Text>
      </View>
      <View style={styles.cellStyle2}>
        <Text>{item.total}</Text>
      </View>
      <View style={styles.cellStyle2}>
        <Text>{item.price}</Text>
      </View>
    </TouchableOpacity>
  );

  const renderEmpty = () => {
    return (
      <View style={styles.empty}>
        <Text>Your inventory is empty</Text>
      </View>
    );
  };

  const TableHeader = () => {
    return (
      <View style={[styles.tableHeader, {backgroundColor: COLORS.primary}]}>
        <View style={styles.cellStyle}>
          <Text style={styles.hText}>Name</Text>
        </View>
        <View style={styles.cellStyle}>
          <Text style={styles.hText}>Desc</Text>
        </View>
        <View style={styles.cellStyle}>
          <Text style={styles.hText}>Total</Text>
        </View>
        <View style={styles.cellStyle}>
          <Text style={styles.hText}>Price</Text>
        </View>
      </View>
    );
  };

  const goToEdit = (name: string) => {
    navigation.navigate('EditInventory', {name});
  };
  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.text} testID="header-text">
          Bluemoon <Text style={styles.inv}>Inventory</Text>
        </Text>
        <TouchableOpacity
          style={styles.addBtn}
          activeOpacity={0.7}
          onPress={goToAddScreen}>
          <AddIcon width={20} height={20} />
        </TouchableOpacity>
      </View>

      <View style={styles.tableContainer}>
        <FlatList
          data={data}
          renderItem={renderItem}
          ListHeaderComponent={TableHeader}
          ListEmptyComponent={renderEmpty}
        />
      </View>

      <CustomButton label="Logout" onPress={logout} />
    </SafeAreaView>
  );
};

export default Home;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
  },
  text: {
    fontWeight: '900',
    fontSize: 30,
  },
  inv: {
    fontWeight: '500',
    fontSize: 16,
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  addBtn: {
    padding: 10,
    backgroundColor: 'white',
    borderRadius: 50,
  },
  tableContainer: {
    marginTop: 20,
    marginBottom: 40,
    flex: 1,
  },

  editButton: {
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'green',
  },
  tableRow: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-evenly',
  },
  tableHeader: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-evenly',
  },
  cellStyle: {
    padding: 7,
    flex: 1,
  },
  cellStyle2: {
    padding: 7,
    flex: 1,
    backgroundColor: 'white',
  },
  hText: {
    fontWeight: '700',
    color: 'white',
  },
  empty: {
    backgroundColor: 'white',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 30,
  },
});
