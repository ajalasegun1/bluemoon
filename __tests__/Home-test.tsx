/**
 * @format
 */

import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
// import renderer from 'react-test-renderer';
import {render, screen} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';

import Home from '../src/screens/Home';

test('Home Screen Snapshot', async () => {
  let props: any;
  //   const tree = renderer.create(<Home {...props} />).toJSON();
  //   expect(tree).toMatchSnapshot();

  const tree = render(
    <NavigationContainer>
      <Home {...props} />
    </NavigationContainer>,
  ).toJSON();
  //   screen.debug();
  expect(tree).toMatchSnapshot();
});
