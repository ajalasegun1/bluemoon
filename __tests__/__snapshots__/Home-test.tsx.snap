// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen Snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "flex": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 30,
          "fontWeight": "900",
        }
      }
    >
      Bluemoon 
      <Text
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "500",
          }
        }
      >
        Inventory
      </Text>
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 50,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <svg
        data-file-name="SvgAdd"
        height={20}
        width={20}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "marginBottom": 40,
        "marginTop": 20,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      ListHeaderComponent={[Function]}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-evenly",
                },
                Object {
                  "backgroundColor": "#279AF1",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "padding": 7,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontWeight": "700",
                  }
                }
              >
                Name
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "padding": 7,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontWeight": "700",
                  }
                }
              >
                Desc
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "padding": 7,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontWeight": "700",
                  }
                }
              >
                Total
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "padding": 7,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontWeight": "700",
                  }
                }
              >
                Price
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "justifyContent": "center",
              "padding": 30,
            }
          }
        >
          <Text>
            Your inventory is empty
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#BFC0C0",
        "borderRadius": 8,
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontWeight": "600",
            "textTransform": "uppercase",
          },
          Object {
            "color": "black",
          },
        ]
      }
    >
      Logout
    </Text>
  </View>
</RNCSafeAreaView>
`;
